### PR TITLE
feat: twingateresourceaccess - allow specifying principal by name

### LINKED ISSUE #32: feat: TwingateResourceAccess allow specifying principal by name
**What is missing?**
In order to get `principalId` developer needs access to Twingate Admin which he, usually doesn't have.

Example:
```
apiVersion: twingate.com/v1beta
kind: TwingateResourceAccess
metadata:
  name: my-twingate-access
spec:
  resourceRef:
    name: my-twingate-resource
    namespace: default
  principalExternalRef:
    type: Group
    matchName: "my group"

```

### PR DESCRIPTION
**Pull request by ekampf**
## Related Tickets & Documents

Resolves #32 

## Changes
Adds support for `principalExternalRef` in the `TwingateResourceAccess` CRD.

```
apiVersion: twingate.com/v1beta
kind: TwingateResourceAccess
metadata:
  name: my-twingate-access
spec:
  resourceRef:
    name: my-twingate-resource
    namespace: default
  principalExternalRef:
    type: Group
    name: "my group"
```

### CODE REVIEW

File: deploy/twingate-operator/crds/twingate.com.twingateresourceaccesses.yaml [hunk 1]
```diff
@@ -16,15 +16,41 @@ spec:
...
               type: object
               description: "TwingateResourceAccessSpec defines the desired state of TwingateResourceAccess"
-              x-kubernetes-validations:
-                - rule: self.principalId == oldSelf.principalId
-                  message: "principalId is immutable"
+              oneOf:
+                - properties:
+                  required: ["resourceRef", "principalId"]
+                - properties:
+                  required: ["resourceRef", "principalExternalRef"]
```
Comment (minhtule) [inline deploy/twingate-operator/crds/twingate.com.twingateresourceaccesses.yaml:23]: Does this prevent us from specifying both `principalId` and `principalExternalRef`? 🤔

File: app/handlers/handlers_resource_access.py [hunk 2]
```diff
@@ -11,6 +11,26 @@
 K8sObject = MutableMapping[Any, Any]
 
 
+def get_principal_id(access_crd: ResourceAccessSpec, client: TwingateAPIClient) -> str:
+    if principal_id := access_crd.principal_id:
+        return principal_id
+
+    if pref := access_crd.principal_external_ref:
```
Comment (minhtule) [inline app/handlers/handlers_resource_access.py:18]: typo: should be `ref`?

File: app/handlers/handlers_resource_access.py [hunk 3]
```diff
@@ -11,6 +11,26 @@
...
+            principal_id = client.get_service_account_id(pref.match_name)
+        else:
+            raise ValueError(f"Unknown principal type: {pref.type}")
+
+        if not principal_id:
+            raise ValueError(f"Principal {pref.type} {pref.match_name} not found.")
+
+        return principal_id
+
+    raise ValueError("missing principal_id or principal_external_ref")
```
Comment (minhtule) [inline app/handlers/handlers_resource_access.py:31]: nit: capital `Missing`

File: app/tests/test_crds_resourceaccess.py [hunk 4]
```diff
@@ -161,6 +161,65 @@ def test_deserialization(sample_resourceaccess_object):
...
+                    "operation": "Update",
+                    "time": "2023-09-29T19:30:39Z",
+                },
+                {
+                    "apiVersion": "twingate.com/v1",
+                    "fieldsType": "FieldsV1",
+                    "fieldsV1": {
+                        "f:spec": {
+                            ".": {},
+                            "f:principalId": {},
```
Comment (minhtule) [inline app/tests/test_crds_resourceaccess.py:194]: Shouldn't this be `principalExternalRef`? Or this `metadata` key is not used in the test? If so, maybe it's simpler to not include it 🤔

**Review by minhtule — APPROVED** [review]
Fix & merge!

File: app/handlers/handlers_resource_access.py (hunk 2 above)
Comment (ekampf) [reply app/handlers/handlers_resource_access.py:18]: pref - principal ref :)

File: deploy/twingate-operator/crds/twingate.com.twingateresourceaccesses.yaml (hunk 1 above)
Comment (ekampf) [reply deploy/twingate-operator/crds/twingate.com.twingateresourceaccesses.yaml:23]: yes.... it says `oneOf`

Comment (ekampf) [reply deploy/twingate-operator/crds/twingate.com.twingateresourceaccesses.yaml:23]: yes.
also added test

**Pull request merged**